[Bug]: Don't retry ResourceNotFoundException in KinesisIO

### What happened?

The not unexpected outcome from #23206 is ResourceNotFoundException from the various AWS SDK calls (if we are hitting the wrong account, the expected stream will not be there). However, current implementation of Kinesis consumer swallows those exceptions and proceeds as if no error had happened, eventually failing with cryptic "NullPointerException" errors (because consumer iterators can not be read/updated properly), like shown below:
```
Caused by: java.lang.NullPointerException
at org.apache.beam.sdk.io.aws2.kinesis.ShardReadersPool.getMinTimestamp(ShardReadersPool.java:257)
at org.apache.beam.sdk.io.aws2.kinesis.ShardReadersPool.getWatermark(ShardReadersPool.java:249)
at org.apache.beam.sdk.io.aws2.kinesis.KinesisReader.getWatermark(KinesisReader.java:133)
at org.apache.beam.runners.flink.translation.wrappers.streaming.io.UnboundedSourceWrapper.onProcessingTime(UnboundedSourceWrapper.java:460)
at org.apache.flink.streaming.runtime.tasks.StreamTask.invokeProcessingTimeCallback(StreamTask.java:1425)
```


### Issue Priority

Priority: 2

### Issue Component

Component: io-java-aws

--- a/sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/SimplifiedKinesisClient.java
+++ b/sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/SimplifiedKinesisClient.java
@@ -34,6 +34,7 @@ import org.joda.time.Instant;
 import org.joda.time.Minutes;
 import software.amazon.awssdk.core.exception.SdkClientException;
 import software.amazon.awssdk.core.exception.SdkServiceException;
+import software.amazon.awssdk.core.internal.retry.SdkDefaultRetrySetting;
 import software.amazon.awssdk.services.cloudwatch.CloudWatchClient;
 import software.amazon.awssdk.services.cloudwatch.model.Datapoint;
 import software.amazon.awssdk.services.cloudwatch.model.Dimension;
@@ -341,12 +342,16 @@ class SimplifiedKinesisClient implements AutoCloseable {
       throw new KinesisClientThrottledException(
           "Too many requests to Kinesis. Wait some time and retry.", e);
     } catch (SdkServiceException e) {
-      throw new TransientKinesisException("Kinesis backend failed. Wait some time and retry.", e);
-    } catch (SdkClientException e) {
-      if (e.retryable()) {
-        throw new TransientKinesisException("Retryable client failure", e);
+      if (e.isThrottlingException()
+          || SdkDefaultRetrySetting.RETRYABLE_STATUS_CODES.contains(e.statusCode())) {
+        throw new TransientKinesisException("Kinesis backend failed. Wait some time and retry.", e);
       }
-      throw new RuntimeException("Not retryable client failure", e);
+      throw e; // others, such as 4xx, are not retryable
+    } catch (SdkClientException e) {
+      if (SdkDefaultRetrySetting.RETRYABLE_EXCEPTIONS.contains(e.getClass())) {
+        throw new TransientKinesisException("Retryable failure", e);
+      }
+      throw e;
     } catch (Exception e) {
       throw new RuntimeException("Unknown kinesis failure, when trying to reach kinesis", e);
     }

--- a/sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/kinesis/SimplifiedKinesisClientTest.java
+++ b/sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/kinesis/SimplifiedKinesisClientTest.java
@@ -42,8 +42,10 @@ import org.mockito.Mock;
 import org.mockito.junit.MockitoJUnitRunner;
 import org.mockito.stubbing.Answer;
 import software.amazon.awssdk.core.SdkBytes;
+import software.amazon.awssdk.core.exception.ApiCallAttemptTimeoutException;
 import software.amazon.awssdk.core.exception.SdkClientException;
 import software.amazon.awssdk.core.exception.SdkServiceException;
+import software.amazon.awssdk.http.HttpStatusCode;
 import software.amazon.awssdk.services.cloudwatch.CloudWatchClient;
 import software.amazon.awssdk.services.cloudwatch.model.Datapoint;
 import software.amazon.awssdk.services.cloudwatch.model.GetMetricStatisticsRequest;
@@ -159,17 +161,20 @@ public class SimplifiedKinesisClientTest {
   @Test
   public void shouldHandleServiceErrorForGetShardIterator() {
     shouldHandleGetShardIteratorError(
-        SdkServiceException.builder().build(), TransientKinesisException.class);
+        SdkServiceException.builder().statusCode(HttpStatusCode.INTERNAL_SERVER_ERROR).build(),
+        TransientKinesisException.class);
   }
 
   @Test
-  public void shouldHandleClientErrorForGetShardIterator() {
-    shouldHandleGetShardIteratorError(SdkClientException.builder().build(), RuntimeException.class);
+  public void shouldHandleRetryableClientErrorForGetShardIterator() {
+    shouldHandleGetShardIteratorError(
+        ApiCallAttemptTimeoutException.builder().build(), TransientKinesisException.class);
   }
 
   @Test
-  public void shouldHandleUnexpectedExceptionForGetShardIterator() {
-    shouldHandleGetShardIteratorError(new NullPointerException(), RuntimeException.class);
+  public void shouldHandleNotRetryableClientErrorForGetShardIterator() {
+    shouldHandleGetShardIteratorError(
+        SdkClientException.builder().build(), SdkClientException.class);
   }
 
   private void shouldHandleGetShardIteratorError(
@@ -505,12 +510,14 @@ public class SimplifiedKinesisClientTest {
   @Test
   public void shouldHandleServiceErrorForShardListing() {
     shouldHandleShardListingError(
-        SdkServiceException.builder().build(), TransientKinesisException.class);
+        SdkServiceException.builder().statusCode(HttpStatusCode.GATEWAY_TIMEOUT).build(),
+        TransientKinesisException.class);
   }
 
   @Test
-  public void shouldHandleClientErrorForShardListing() {
-    shouldHandleShardListingError(SdkClientException.builder().build(), RuntimeException.class);
+  public void shouldHandleRetryableClientErrorForShardListing() {
+    shouldHandleShardListingError(
+        ApiCallAttemptTimeoutException.builder().build(), TransientKinesisException.class);
   }
 
   @Test
@@ -602,17 +609,20 @@ public class SimplifiedKinesisClientTest {
   @Test
   public void shouldHandleServiceErrorForGetBacklogBytes() {
     shouldHandleGetBacklogBytesError(
-        SdkServiceException.builder().build(), TransientKinesisException.class);
+        SdkServiceException.builder().statusCode(HttpStatusCode.SERVICE_UNAVAILABLE).build(),
+        TransientKinesisException.class);
   }
 
   @Test
-  public void shouldHandleClientErrorForGetBacklogBytes() {
-    shouldHandleGetBacklogBytesError(SdkClientException.builder().build(), RuntimeException.class);
+  public void shouldHandleRetryableClientErrorForGetBacklogBytes() {
+    shouldHandleGetBacklogBytesError(
+        ApiCallAttemptTimeoutException.builder().build(), TransientKinesisException.class);
   }
 
   @Test
-  public void shouldHandleUnexpectedExceptionForGetBacklogBytes() {
-    shouldHandleGetBacklogBytesError(new NullPointerException(), RuntimeException.class);
+  public void shouldHandleNotRetryableClientErrorForGetBacklogBytes() {
+    shouldHandleGetBacklogBytesError(
+        SdkClientException.builder().build(), SdkClientException.class);
   }
 
   private void shouldHandleGetBacklogBytesError(

